### ISSUE TITLE
Ended Zippers

### ISSUE THREAD

**Issue by novaugust** [doorgan/sourceror#67]
I had code that was going to reduce a series of transformations over a zipper, written something like this:
```ex
transformations
|> Enum.reduce(Zipper.zip(ast), fn transformation, zipper ->
  zipper
  |> Zipper.traverse(&transformation.(&1))
  |> Zipper.top()
end)
|> Zipper.root()
```

but only the first transformation is ever run, because `Zipper.top` and all other functions essentially noop on `{_tree, :end}`.
I realize now that when `traverse` is done, the node is always the top of the tree so there's no need to be calling `top` afterwards regardless, and that really the issue you can't chain-traverse is the `:end` atom itself.

It seems like the `:end` atom is an implementation detail of `traverse` that has leaked its way to the rest of the functions. would it be incorrect to change the implementation of `do_traverse` like so?

```diff
-  defp do_traverse({tree, :end}, _), do: {tree, :end}
+  defp do_traverse({tree, :end}, _), do: {tree, nil}
```

this would allow a tree to be repeatedly traversed, and remove the need to handle `:end` as an equivalent to nil in so many functions as added in 35c8714b918b2bea9d836c0163024eda5ea50248

the only other function that can result in an end tuple are next_up and prev_up helpers
```ex
  defp next_up(zipper) do
    parent = up(zipper)

    if parent do
      right(parent) || next_up(parent)
    else
      {node(zipper), :end}
    end
  end

  defp prev_up(zipper) do
    parent = up(zipper)

    if parent do
      left(parent) || prev_up(parent)
    else
      {node(zipper), :end}
    end
  end
```
each of which could return an error (or nil) instead of the :end tuple, removing the special case of `:end` from the (exposed) type entirely - and letting me run traversals without having to wrap/unwrap or dig into the data ;)

if the idea seems solid to you i'd be happy to PR. thanks as always for the great library and all the fun things you've taught me.

**Comment by doorgan** [comment]
I think that's actually intentional, as you don't want to endlessly recurse a zipper with `next`, the concept of an ended zipper serves this purpose so other functions need to be aware of this. However I agree that there should be a way to let you re-traverse a zipper.

I consider the zipper `meta` to be an opaque data structure so while you could in theory just replace `:end` with `nil` in your zipper to continue traversing, I think it would be Sourceror's responsibility to do that.

So instead of removing the idea of an ended zipper, I'd instead introduce something like `reset/1`:
```elixir
def reset({tree, :end}), do: {tree, nil}
def reset(zipper), do: zipper
```

So your code would instead be:
```elixir
transformations
|> Enum.reduce(Zipper.zip(ast), fn transformation, zipper ->
  zipper
  |> Zipper.traverse(&transformation.(&1))
  |> Zipper.reset()
end)
|> Zipper.root()
```

What do you think?

**Comment by novaugust** [comment]
> I consider the zipper `meta` to be an opaque data structure so while you could in theory just replace `:end` with `nil` in your zipper to continue traversing, I think it would be Sourceror's responsibility to do that.

agreed, that's how i like handling it as well =) it would indeed be trivial to just swap that second element and loop again, but it feels like going too deep under the hood

>  as you don't want to endlessly recurse a zipper with next

if next returned `nil` (or another error) when there was no next you'd still be safe from endlessly recursing:

```diff
# def next_up
    ...
    if parent do
      right(parent) || next_up(parent)
-    else
-    {node(zipper), :end}
    end
```
essentially, `nil` could be used to encode illegal moves (like it is in some places, like `right/1`), rather than returning `{where_you_were, :end}` for illegal moves
all that said, i realize i'm talking breaking changes so no stress

>So instead of removing the idea of an ended zipper, I'd instead introduce something like reset/1:

that would solve the problem, but rather than expanding the api, what about changing `top/1` to have the behavior you're proposing so that the original code works? semantically "go to the top" to mean means you've reset the tree anyways, and are no longer at the "end" =)

but if reset is the way to go for ease of introduction, what do you think of this definition?

```ex
def reset({tree, :end}), do: {tree, nil}
def reset(zipper), do: top(zipper)
```

thanks for the discussion!

**Comment by novaugust** [comment]
> essentially, `nil` could be used to encode illegal moves (like it is in some places, like `right/1`), rather than returning `{where_you_were, :end}` for illegal moves

admittedly, this'll result in a lot of `if next = next(zipper), do: recurse(next), else: zipper` style clauses, but we can see that the `:end` tuple similarly results in lots of `def recurse({_, :end} = zipper), do: zipper` clauses so it'd probably have about the same footprint

(sorry, this extra message is me thinking out loud ^.^)

**Comment by novaugust** [comment]
Yeaah now that it's the weekend and i've looked more closely, sure seems like removing the concept of `:end` is a very big api change :) So I'd go for adding reset or changing how top works as a solution to the re-traversal problem. Thanks doorgan

**Comment by novaugust** [comment]
and just for kicks (there's a blizzard here this weekend, so nothing else to do), here's what endless zippers would look like:
https://github.com/novaugust/sourceror/commit/d7a0a6d462ec77262b07f86cdeeba64937d4ff4e

**Comment by doorgan** [comment]
After looking at how the projects using sourceror are using zippers, I'm not that concerned about this being a breaking change, I was more concerned about the semantics of the zipper itself. This module is largely based on the clojure/zip implementation of zippers, so the "ended zipper" part came from there.

Your solution to remove ended zippers seems to work, so a PR would be welcome!
I'd like to give a heads up to @NickNeck before going this route though, because the `zipper_ex` if memory serves is based on Sourceror's implementation of zippers.

**Comment by doorgan** [comment]
Closed in #68
